design_documents/transactional_testing/transactional_testing.md

```diff
@@ -0,0 +1,172 @@
+# Transactional testing proposal
+
+## Context
+
+Some flows are not currently intuitively testable via standard testing methods, 
+like accounts logic (__validate__, __execute__ etc.), transaction rejections,
+fees, contract upgrade flows, and other e2e testing procedures.
+
+## Goal
+
+In order to facilitate that, we should consider adding a way to test
+the contracts in conditions as close to real networks conditions as possible.
+
+Also, the goal is to make the deployments testable and validation 
+of the deployments possible to the user. 
+
+The tests should be also runnable in a variety of environments, depending on the
+use case (namely katana, devnet, or testnet). This would increase user's confidence
+in the release process.
+
+## Proposed solution
+
+We could consider extending `cast script`, since the functionalities would overlap.
+This could be considered as a variant of `cast script`, in a way. 
+
+The extending/differentiating factors would be:
+
+- Extra environment-specific functions (for [katana](https://book.dojoengine.org/toolchain/katana/reference.html#custom-methods)/[devnet](https://github.com/0xSpaceShard/starknet-devnet-rs#dumping--loading) - see docs)
+- Additional RPC functions support (receipts, etc.)
+- Test-like behavior (fail/pass)
+- Idempotent functions, for functionalities that can fail when double-running them (i.e. declare)
+
+
+## Solution architecture
+
+### High-level architecture
+This is pretty straightforward. We ought to use the existing architecture for hint interception,
+provide extra functionality via libraries (one for katana, second one for devnet, possibly more), and collect the cases from
+folders using existing collecting logic. 
+
+Note: Currently, the backend architecture between `cast script` and `forge test` is not common,
+and the following effort to implement this architecture should be preceded by making that solution.
+
+![txn_based_testing_arch.png](./txn_based_testing_arch.png)
+
+### E2E tests and scarb
+
+The relation between scarb and the following tests is very much like in `cast script`:
+1. Scarb compiles the contracts
+2. The test scripts are compiled with foundry (until a plugin solution is available)
+3. The test uses artifacts from Ad. 1, runs program from Ad 2. on own VM with hint processing logic 
+
+### File organization within scarb workspace
+
+Each package can have its own contract, so each package could have its own e2e directory, 
+and the whole workspace could have it as well (to test it in integration).
+For example:
+```
+./project
+  ./Scarb.toml
+  ./e2e
+    ./e2e/test_integration.cairo
+    ./e2e/test_integration2.cairo
+  ./pkg1
+    ./src
+    ./test
+    ./e2e
+      ./test_contract_1.cairo
+  ./pkg2
+  ...
+```
```

Comment:
I wonder if theses shouldn't go under `test/e2e`, though it would be harder to maintain.

Reply:
I would prefer not to change the collection logic on forge's side once again :/ In theory tests under `tests/e2e` won't be collected if there is no `tests/e2e.cairo` file but we there is nothing preventing someone from having forge tests structured like this:

```
tests/
  e2e.cairo   <- # mod my_forge_test; #
  e2e/
     my_forge_test.cairo
```